### PR TITLE
Working AES-CBC using the CAU on the Teensy 3

### PR DESCRIPTION
**Pull request by jeremy-daily**

### CODE REVIEW

**Review by duyvan1995 — APPROVED** [review]

File: tests/cryptolibAESSHA/cau_api.h [hunk 1]
```diff
@@ -0,0 +1,390 @@
+/*
```
Comment (BenGardiner) [inline tests/cryptolibAESSHA/cau_api.h:1]: is this *your* header file? did you import it from somewhere else? If so, how is it licensed?

File: tests/cryptolibAESSHA/cryptolibAESSHA.ino [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+void aes_cbc_encrypt(const unsigned char *data, unsigned char *cipher_text){
+  //Data length should be a multiple of 16 bytes
+  // Need to initialize "out" with the initialization vector
+  for (uint32_t j=0; j < data_size; j+=16){
+    for (uint8_t i = 0; i < 16; i++){
+      in[i] = data[j+i] ^ out[i];
+    }
+    cau_aes_encrypt (in, keysched, 10, out); // # 16-byte block
```
Comment (BenGardiner) [inline tests/cryptolibAESSHA/cryptolibAESSHA.ino]: magic number `10` -- I recommend defining this as a descriptive-named constant. e.g. AES_128_NROUNDS

File: tests/cryptolibAESSHA/cryptolibAESSHA.ino [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  } Serial.println();
+
+   
+  sprintf(str, "aes cbc encrypt %d bytes %u us, KBs ", sizeof(data), t);
+  Serial.print(str);
+  Serial.println(1000.*sizeof(data) / t);
+  
+  t = micros();
+  memcpy(iv,init_vector,16);
+  aes_cbc_decrypt(clear_text, cipher_text);
```
Comment (BenGardiner) [inline tests/cryptolibAESSHA/cryptolibAESSHA.ino:121]: you're testing 'round-trip' through your aes-cbc routines; this is a good test.

It is necessary but not sufficient. You should also test with known test vectors. e.g. https://github.com/google/wycheproof/tree/master/testvectors

File: tests/cryptolibAESSHA/mmcau_des_functions.S [hunk 4]
```diff
@@ -0,0 +1,257 @@
+#*******************************************************************************
+#*******************************************************************************
+#
+# Copyright (c) Freescale Semiconductor, Inc 2011.
```
Comment (BenGardiner) [inline tests/cryptolibAESSHA/mmcau_des_functions.S:4]: under what license does Freescale distribute this assembler file?

File: tests/cryptolibAESSHA/mmcau_aes_functions.S [hunk 5]
```diff
@@ -0,0 +1,929 @@
+#*******************************************************************************
+#*******************************************************************************
+#
+# Copyright (c) Freescale Semiconductor, Inc 2011.
```
Comment (BenGardiner) [inline tests/cryptolibAESSHA/mmcau_aes_functions.S:4]: under what license does Freescale distribute this assembler file?

File: tests/cryptolibAESSHA/mmcau_des_functions.S (hunk 4 above)
Comment (BenGardiner) [reply tests/cryptolibAESSHA/mmcau_des_functions.S:4]: same question for the remaining `.S` files too

**Review by BenGardiner — CHANGES_REQUESTED** [review]

File: tests/cryptolibAESSHA/cau_api.h (hunk 1 above)
Comment (jeremy-daily) [reply tests/cryptolibAESSHA/cau_api.h:1]: This is from Freescale. I've included the details on the license from their download.

File: tests/cryptolibAESSHA/cryptolibAESSHA.ino (hunk 3 above)
Comment (jeremy-daily) [reply tests/cryptolibAESSHA/cryptolibAESSHA.ino:121]: Looks interesting. I'll have to look into this more. It looks challenging to port...

File: tests/cryptolibAESSHA/mmcau_des_functions.S (hunk 4 above)
Comment (jeremy-daily) [reply tests/cryptolibAESSHA/mmcau_des_functions.S:4]: Posted the license. It's permissive.

File: tests/cryptolibAESSHA/mmcau_aes_functions.S (hunk 5 above)
Comment (jeremy-daily) [reply tests/cryptolibAESSHA/mmcau_aes_functions.S:4]: Posted the license. It's permissive.

File: tests/cryptolibAESSHA/cryptolibAESSHA.ino (hunk 2 above)
Comment (jeremy-daily) [reply tests/cryptolibAESSHA/cryptolibAESSHA.ino]: Done.

File: tests/cryptolibAESSHA/cryptolibAESSHA.ino (hunk 3 above)
Comment (BenGardiner) [reply tests/cryptolibAESSHA/cryptolibAESSHA.ino:121]: No need to port the project; but you should include tests that confirm that your implementation here is capabale of encrypting a known plaintext into and expected ciphertext and the analogous test for decrypt.  Right now your tests are only confirming that you have a pair of functions that are a matching encrypt/decrypt; but you're not confirming that existing tools/implementations could encrypt or decrypt the outputs of your implementations

File: tests/cryptolibAESSHA/cryptolibAESSHA.ino [hunk 6]
```diff
@@ -4,7 +4,9 @@ Uses 512 Byte block examples
 */
 #include "CryptoAccel.h" //Makes it arduino compatible
 #include "cau_api.h"
+
 #define data_size 512
+#define AES_ROUNDS 10
```
Comment (BenGardiner) [inline tests/cryptolibAESSHA/cryptolibAESSHA.ino]: not quite descriptive enough; This is a *value* for the number of AES rounds, true; however, for AES other than AES-128 it is not a valid number of rounds...

File: tests/cryptolibAESSHA/cryptolibAESSHA.ino (hunk 3 above)
Comment (jeremy-daily) [reply tests/cryptolibAESSHA/cryptolibAESSHA.ino:121]: We tested with the online calculator at https://www.cryptogrium.com/aes-cbc.html and obtained matching results.

File: tests/cryptolibAESSHA/cryptolibAESSHA.ino (hunk 6 above)
Comment (jeremy-daily) [reply tests/cryptolibAESSHA/cryptolibAESSHA.ino]: Easy change. I replaced the one @dcn1470 entered too.

File: tests/cryptolibAESSHA/cryptolibAESSHA.ino (hunk 3 above)
Comment (BenGardiner) [reply tests/cryptolibAESSHA/cryptolibAESSHA.ino:121]: You should use the encrypted output of that online calculator as the input in a test here (and vice versa). Right now your tests verify that your implementation can do a round trip. You are missing a test that to confirm that the implementation will 'work with standard tools'.

**Review by BenGardiner — APPROVED** [review]

**Comment by BenGardiner** [comment]
I still think you should have tests against static/constant/literal test-vectors and not just round-trip your own implementations. But PR #6 is on-top-of this one and I want to review it separately.

**Comment by jeremy-daily** [comment]
Agreed. I think we should start a new branch with the test vectors.

**Pull request merged**